Create a docker multistage build script

dkuegler opened the pull request:
Create a docker multistage build script that is compatible with FastSurfer interface rework.

All images are built with the "core" Dockerfile at Docker/Dockerfile. Also update some additional folders to not be copied automatically by the build (.dockerignore).

TODOs:
- Testing
- check for the updates to AMD (rocm/pytorch is now python 3.8 as well)
-

dkuegler committed:
Create a docker multistage build script that is compatible with FastSurfer interface rework.

All images are built with the "core" Dockerfile at Docker/Dockerfile.
Also update some additional folders to not be copied automatically by the build (.dockerignore).

dkuegler commented:
@AhmedFaisal95 Please have a look and see what still needs to be done.
I have already added some documentation in Docker/README.md

At a minimum, this drastically reduces the doubling of code. 

It seems that DOCKER_BUILDKIT also supports the script now. Which speeds up builds a lot, since it builds stages that are independent in parallel.

For example
```
DOCKER_BUILDKIT=1 docker build --rm=true --target runtime --build-arg DEVICE=cuda --build-arg FREESURFER=pruned -t fastsurfer:gpu -f ./Docker/Dockerfile .
```

af-a commented:
This seems to run fine except when:
1. build arg `DEVICE` is set to `cpu` (any target) 
2. the target is `runtime_surf_only` and `DEVICE` is set to `none`.

### Case 1
Command:
```
DOCKER_BUILDKIT=1 docker build --rm=true --target runtime --build-arg DEVICE=cpu -t IMAGE_ID -f ./Docker/Dockerfile .
```

Result:
```
[+] Building 0.6s (11/17)                                                                                                                                                                                                                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                                                           0.3s
 => => transferring dockerfile: 5.62kB                                                                                                                                                                                                                                                                                                                                         0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                              0.4s
 => => transferring context: 136B                                                                                                                                                                                                                                                                                                                                              0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                                                                                                                                                                0.0s
 => CANCELED [internal] load build context                                                                                                                                                                                                                                                                                                                                     0.0s
 => [build_conda_base 1/3] FROM docker.io/library/ubuntu:20.04                                                                                                                                                                                                                                                                                                                 0.0s
 => CACHED [build_conda_base 2/3] RUN apt-get update && apt-get install -y --no-install-recommends       wget       git       ca-certificates       upx       file &&     apt clean &&     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*                                                                                                                                       0.0s
 => CACHED [build_conda_base 3/3] RUN wget --no-check-certificate -qO ~/miniconda.sh https://repo.continuum.io/miniconda/Miniconda3-py38_4.11.0-Linux-x86_64.sh  &&      chmod +x ~/miniconda.sh &&      ~/miniconda.sh -b -p /opt/conda &&      rm ~/miniconda.sh                                                                                                             0.0s
 => CACHED [build_conda_cpu 1/1] RUN conda install --channel pytorch cpuonly=1.0                                                                                                                                                                                                                                                                                               0.0s
 => CACHED [runtime 1/7] RUN apt-get update && apt-get install -y --no-install-recommends       tcsh       time       bc       gawk       libgomp1       libquadmath0 &&     apt clean &&     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*                                                                                                                                    0.0s
 => CACHED [runtime 2/7] RUN echo "source /venv/bin/activate" >> ~/.bashrc                                                                                                                                                                                                                                                                                                     0.0s
 => ERROR [runtime 3/7] COPY --from=build_conda /venv /venv                                                                                                                                                                                                                                                                                                                    0.0s
------
 > [runtime 3/7] COPY --from=build_conda /venv /venv:
------
failed to compute cache key: "/venv" not found: not found
```


### Case 2
Command:
```
DOCKER_BUILDKIT=1 docker build --rm=true --target runtime_surf_only --build-arg DEVICE=none --build-arg FREESURFER=pruned -t IMAGE_ID -f ./Docker/Dockerfile .
```

Result:
```
[+] Building 1.6s (4/4) FINISHED                                                           
 => [internal] load build definition from Dockerfile                                  0.3s
 => => transferring dockerfile: 140B                                                  0.0s
 => [internal] load .dockerignore                                                     0.5s
 => => transferring context: 136B                                                     0.0s
 => ERROR [internal] load metadata for docker.io/library/base_none:latest             1.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                       0.0s
------
 > [internal] load metadata for docker.io/library/base_none:latest:
------

```

af-a committed:
[Dockerfile] Add base_none base image, fix build_conda_cpu

This enables building runtime_surf_only with DEVICE=none.
Fix: Change build_conda_cpu base (build_conda_base --> build_conda_common)

af-a commented:
The issues mentioned above should be fixed in this PR: https://github.com/dkuegler/FastSurfer/pull/2

dkuegler commented:
@AhmedFaisal95 Can you confirm this works now?

We need a "basic test" of all images :).

dkuegler commented:
@m-reuter The Apple M1 device is missing from the docker multistage built. The reason is that it was not in `dev` at all.

For the AMD build, the base image got updated and I am not really sure if the updated image works. Maybe, someone can check this as well?

af-a commented:
@dkuegler 

> @AhmedFaisal95 Can you confirm this works now?
> 
> We need a "basic test" of all images :).

I already tested the fixes before submitting the PR, specifically all images except for the AMD case (no hardware).

dkuegler committed:
Add the MPLCONFIGDIR VARIABLE to the base docker container

dkuegler commented:
The build script should also compile the byte code https://docs.python.org/3/library/py_compile.html .

dkuegler committed:
Compile fastsurfer python code into bytecode.

Add some more files and folders to .dockerignore
Remove the chmod on checkpoints command from the Dockerfile

dkuegler commented:
I think what is left for this PR to be merged is proper testing.
Especially for the AMD build.

m-reuter commented:
> The Apple M1 device is missing from the docker multistage built. The reason is that it was not in `dev` at all.

For M1 we do not use a docker, only native install on Mac. Inside Docker it was much slower for yet unknown reasons, see #109

dkuegler commented:
@agirodi can you test the AMD docker script?

agirodi commented:
> @agirodi can you test the AMD docker script?

I just did and it is working correctly. The build went without issues and running produced a correct segmentation.

dkuegler commented:
Then, this PR should be ready for merging.

m-reuter commented:
A few points for documentation:
- due to the minimal larger size when including FreeSurfer binaries, we will stop publishing docker images without FreeSurfer
- pytorch gpu is still much larger than cpu version so we will continue to publish both versions
- AMD is experimental and can break without notice, which is why we will provide build files, but will not publish the container for now. 
- As dependencies (versions) remain relatively stable, we will have a base release image without pytorch (including all other dependencies like python packages and FreeSurfer binaries), and then two versions from that with pytorch GPU and CPU. 
- FastSurfer is then added on top of these three as the final layer. The non-pyorch entry point remains recon-surf, for the other two it will be run_fastsurfer.sh.. Users who want to directly run_prediction, need to modify the docker run command and overwrite the entry point or add another layer. 

The advantage of this approach is, that for future docker builds we can pull the base release images and use them (in most cases they probably exist locally anyway). Adding FastSurfer is then extremely fast :-)

dkuegler committed:
Modify Scripts so build images can be easily reintegrated into new builds

dkuegler committed:
Dockerfile documentation

dkuegler committed:
port the docker build script to python
change the file with BUILD information to BUILD.txt (instead of VERSION.txt)
change the formatting options in segstats.py

dkuegler committed:
Update Docker README documentation

dkuegler committed:
A lot of changes to the multistage docker scripts

Fundamentally, the cpu image now is ~3GB and does not include the cudatoolkit any more.
pytorch is now installed from the pytorch download page with all relevant platform-specific requirements such as the cudatoolkit or rocm drivers/runtime.
All this now always part of the conda environment.

- Update/fix build.py build script for python
- add conda-pack script to pack the conda environment
- update the Docker/README.md
- added an automatically created build.log file into each subject directory
- added a BUILD.info file into the docker container, that contains version information about version number, branch, git status at build time, checkpoints contained in container, and pip package list.
- added a FastSurferCNN/version.py script that replaces the version-related commands in run_fastsurfer.sh
- added baked in support for AWS
- remove torchio as standard imports in FastSurferCNN/data_loader/(dataset|loader).py, so we do not need torchio at inference time any more.
- added FastSurferCNN/utils/run_tools, a wrapper to subprocess.Popen that allows easier, non-blocking executiong of scripts and binaries.
- added BUILD.info to .gitignore
- removed BUILD.txt from .dockerignore
- run_fastsurfer.sh now also supports a --version command, see --help
- deactivate user-site imports (-s flag to python)

dkuegler committed:
- cleanup of the installation directory in the docker image
- improve debug info of image (in /install folder, if --build-arg DEBUG=true)
- add --progress=plain to docker build if running inside the python script because no over-write on the console is possible
- merge different pip sections in the Dockerfile, as multiple sections were not merged automatically by conda
- filter __pycache__ from version
- up ubuntu default versions to 22.04
- fix python -s for no user directory packages

m-reuter commented on Docker/Dockerfile:
```diff
@@ -1,102 +1,201 @@
 
-# Install required packages for freesurfer to run
+ENV LANG=C.UTF-8 \
+    HSA_OVERRIDE_GFX_VERSION=10.3.0
```
maybe we need a comment here, why this is needed? And when it would need to be changed to a new version?

dkuegler replied:
This is a flag for AMD ROCm. It should probably not be part of the "standard" docker.
In fact, since this is a flag that enables ROCm support on hardware that AMD does not officially support ROCm for, it should probably not be part of the docker container at all.
I suggest this flag **has to be passed** by the user `docker run ... -e HSA_OVERRIDE_GFX_VERSION=10.3.0 fastsurfer:rocm ...` instead of blanket-setting it here. Because other hardware might need this variable to be "unset".

m-reuter replied:
yes, makes sense. It should be added to the documentation (README). I think our AMD GPU needed this flag and probably other consumer cards.

m-reuter commented on Docker/README.md:
```diff
@@ -42,35 +43,42 @@ All other available flags are identical to the ones explained on the main page [
                       -v /home/user/my_fastsurfer_analysis:/output \
                       -v /home/user/my_fs_license_dir:/fs_license \
-                      --rm --user $(id -u):$(id -g) my_fastsurfer:gpu \
+                      --rm --user $(id -u):$(id -g) my_fastsurfer:2.2.1 \
```
We did not want version numbers hard coded in the documentation.

m-reuter commented on Docker/README.md:
```diff
@@ -83,16 +91,15 @@ docker run --gpus all -v /home/user/my_mri_data:/data \
            -v /home/user/my_fastsurfer_analysis:/output \
            -v /home/user/my_fs_license_dir:/fs_license \
-           --rm --user $(id -u):$(id -g) my_fastsurfer:cpu \
+           --rm --user $(id -u):$(id -g) my_fastsurfer:2.2.1-cpu \
```
another hard coded version

m-reuter commented on Docker/README.md:
```diff
@@ -191,7 +128,7 @@ docker run --rm --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
 	   -v /home/user/my_mri_data:/data \
 	   -v /home/user/my_fastsurfer_analysis:/output \
-	   my_fastsurfer:gpu-amd \
+	   my_fastsurfer:2.2.1-rocm \
```
another hard coded version

dkuegler committed:
- remove version numbers in README.md
- add `--tag my_fastsurfer:<device>` in README.md
- remove rocm environment variable
- add documentation for HSA_OVERRIDE_GFX_VERSION
